### ISSUE TITLE
Prevent ORA-01000 when executing multiple insert statements

### ISSUE THREAD

**Issue by somax80** [ADOdb/ADOdb#770]
Hi,
i'm getting an Oracle error "ORA-01000: maximum open cursors exceeded" when executing an insert statement (prepared with "Prepare" and with binding parameters).
The sql statement is into a  for cycle and i've got the error when the numbers of iterations (that equals the numbers of insert statements), reaches the limit of open cursors set on DB server side (open_cursors setting).
With oci functions i could free the statement after each Execute, but in adodb there isn't any method to free the statement, so i couldn't release the cursor that was automatically opened in Oracle for every statement.

This is a basic sample code to reproduce the error (it use a sample oracle table with two varchar2(50) fields):
![example](https://user-images.githubusercontent.com/82518160/142213599-ea434a78-c418-4ca6-bea4-11f04af6f2b1.jpg)

How i could explicitly free the statement?
I don't found any method to do this and i don't have any valid recordset to call _close method after execute.

I use:
PHP 7.4.19
OCI8 Version 2.2.0
Oracle client 19.11.0.0.0
Oracle server 19.11.0.0.0
AdoDb library version 5.21.0

Thanks,
  Max

**Comment by mnewnham** [comment]
I am reviewing this but it may take me a day or so to respond

**Comment by mnewnham** [comment]
Can I suggest that you use the ultra-fast **bulk bind** method.  You must start by enabling bulkbinding in the driver

````php
/*
* Enable bulk binding
*/
$db->bulkBind = true;
````
Then you need to restructure your code to start with a 2 dimensional array of values

````php

$bindVars = array();
for ($i=0;$i<500;$i++)
{
  $bindVars[] = array('Field1' . $i,'Field2' . $i);
}
````
Next pass the array directly to the **execute()**
````php
$sql = "INSERT INTO TBLORA001(FIELD1,FIELD2) values (:pf0,:pf1)";
$db->execute($sql,$bindVars)
````

**Comment by somax80** [comment]
Thanks for the suggestion. I try to modify my sample with "bulkbind" and i haven't got any runtime error.

But, if for some reason I MUST use one execute for any insert/update (like my sample), how could I free the statement after each execute?
An use case may be if i need to run ten different insert/update sequentially releasing the statement_id after any execute,
Another possible use case is if i have the execution of single row insert statement in a method and i would call Ntimes this method in a cycle, passing the connection as parameter.

If I rewrite my example code using oci functions, after any oci_execute($stmt) I could simply call oci_free_statement($stmt) to release the statement and i haven't got any error.
But in AdoDb (used in all my applications) i don't find any method to release the statement. I find some calls to oci_free_statement embedded in other methods, but only in certain conditions.
How is re-mapped oci_free_statement function in AdoDb?

Thanks,
  Max

![examaple](https://user-images.githubusercontent.com/82518160/142585445-3512ec14-0877-4d39-81be-d0243aec750d.jpg)

**Comment by dregad** [comment]
Since you're using a prepared statement, I'm wondering why are you preparing it over and over again with each iteration ? The point of having a prepared statement is that you can execute it multiple times with different parameters, so

```
$stmt = $conn->prepare($sql);
loop {
   $conn->inParameter(...);
   $conn->execute($stmt);
}
```

As a side note, instead of pasting a screenshot of your code, I suggest you post the actual code (text) surrounded by ` ``` ` tags

**Comment by somax80** [comment]
Ok, but consider this case: in an application there is a class called "User" with a method "modify" that incapsulate the sql string, prepare+bind+execute of the update statement to modify ONE user. The statement could be one ore more and could be different based on a property of the user.

If i want to modify more than one "user", i've an array of "User" instances ad i call the method "modify" for each instance, in a cycle, passing to the method only the db connection identifier.
In this case i need to put into the method "modify" also the release of the statement, to avoid ORA-01000 exception.

Basically the question is:
Is there a way to manually release an OCI statement hold by adodb?
Is there a method that wrap oci_free_statement function in AdoDb?

Many thanks
  Max

**Comment by dregad** [comment]
> Is there a method that wrap oci_free_statement function in AdoDb?

Not currently.

> Is there a way to manually release an OCI statement hold by adodb?

At first I thought, considering that the query is not being reused, that you could simply call `$conn->execute($sql, $par);` but looking at the code, I'm afraid it would not release the statement either.

ADOdb does not actually have a concept of statement objects, we don't "hold" them.

But the _oci8 statement_ resource is given to you in the array returned by prepare() method, so if portability is not an issue I guess you could simply release it with `oci_free_statement($stmt[1]);` or `$stmt[1] = null;` as explained in [oci_parse() function documentation](https://www.php.net/manual/en/function.oci-parse.php)

**Comment by mnewnham** [comment]
Just reviewing your original query, I think in fact there is no need to use prepare at all, the query works correctly like this:

````php

$bindVars = array('pf0'=>'Any Value','pf1'=>'Some other value');

$sql = "INSERT INTO TBLORA001(FIELD1,FIELD2) values (:pf0,:pf1)";
$db->execute($sql,$bindVars);
````
The standard execute for Oracle automatically uses features such as oci_parse for the SQL statement and oci_bind_by_name for the bind variables. [InParameter](http://adodb.org/dokuwiki/doku.php?id=v5:reference:connection:inparameter()) is used primarily with [prepareSp()](http://adodb.org/dokuwiki/doku.php?id=v5:reference:connection:preparesp) for sending parameters to stored procedures

**Comment by somax80** [comment]
Thanks, i've modified the code without using prepare and inparameter:
```
$conn = NewADOConnection('oci8://user:password@myalias/?nls_date_format=DD/MM/YYYY');
$conn->debug=false;
$conn->BeginTrans();
for ($i=0;$i<500;$i++) {
  $par=array('pf1'=>'Field1.'.$i,
			     'pf2'=>'Field2.'.$i);
  $sql_string= "insert into TBLORA10000 (FIELD1, FIELD2) values (:pf1,:pf2)";
  //$stmt = $conn->Prepare($sql_string);
  $rs=$conn->Execute($sql_string,$par);
  //oci_free_statement($stmt[1]);
}
$conn->CommitTrans();
$conn->Close();
```
The script run successfully, but i've got the same error ORA-01000 when the insert statements raise max open cursors value.
So i must use prepare statement, because i need the statement ID to call oci_free_statement on $stmt[1] on every iteration.

If it matters, i found that the problem of ORA-01000 occurs also with calling directly oci functions (without using AdoDb):
```
$conn = oci_connect('user','password','myalias');
for ($i=0;$i<500;$i++) {
  $par=array('f1'=>'Field1.'.$i,
			     'f2'=>'Field2.'.$i);
  $sql_string= "insert into TBLORA10000 (FIELD1, FIELD2) values (:pf1,:pf2)";	
  $stmt = oci_parse($conn,$sql_string);
  oci_bind_by_name($stmt, ":pf1", $par['f1']);
  oci_bind_by_name($stmt, ":pf2", $par['f2']);
  $rs=oci_execute($stmt,OCI_NO_AUTO_COMMIT);
}
oci_commit($conn);
oci_close($conn);
```
With PHP 7.4.19, OCI8 Version 2.2.0, Oracle client 19.11.0.0.0, Oracle server 19.11.0.0.0 the problem OCCURS.
With PHP 5.6.40, OCI8 Version 2.0.10, Oracle client 19.11.0.0.0, Oracle server 19.11.0.0.0  it will automatically release the statement, so the problem DOESN'T OCCURS.
So, the problem is seems not related to AdoDb, but to be dependent on PHP version or OCI8 PHP module version.
All others component are the same, also PHP, OCI8 and Oracle configuration.

In my opinion, if i could, i suggest you to implement the free statement functionality in next release of AdoDb oci8 driver.

Thanks to all,
best regards
Max

**Comment by somax80** [comment]
If anyone needs it, i've created an example that extended oci8 driver to implements a new method "ReleaseConnection", that free the statements returned in array from Prepare.

```
require_once 'adodb/adodb.inc.php'; 
require_once 'adodb/drivers/adodb-oci8.inc.php'; 

class My_oci8 extends ADODB_oci8 {  
 
  var $rsPrefix = 'My_rs_';  
 
  function ReleaseStatement(&$stmt) {
	  if (!is_array($stmt)) {
			return false;
		}
		
	if (is_resource($stmt[1])) {
		oci_free_statement($stmt[1]);
		return true;
	}
	
	return false;
  }  
}
 
class My_rs_oci8 extends ADORecordSet_oci8 {   
}  
  
 
$ADODB_NEWCONNECTION = 'My_oci8_factory';  
 
function& My_oci8_factory($driver) {  
         if ($driver !== 'oci8') return false;  
         $driver = 'My_'.$driver;  
         $obj = new $driver();  
         return $obj;  
}

...
$conn = NewADOConnection('oci8://user:password@myalias/?nls_date_format=DD/MM/YYYY');
...
$par=array('pf1'=>'Field1.'.$i,
			   'pf2'=>'Field2.'.$i);
$sql_string= "insert into TBLORA10000 (FIELD1, FIELD2) values (:pf1,:pf2)";
$stmt = $conn->Prepare($sql_string);
$rs=$conn->Execute($stmt,$par);
$conn->ReleaseStatement($stmt);
```

best regards, 
  Massimo

**Comment by somax80** [comment]
> With PHP 7.4.19, OCI8 Version 2.2.0, Oracle client 19.11.0.0.0, Oracle server 19.11.0.0.0 the problem OCCURS. With PHP 5.6.40, OCI8 Version 2.0.10, Oracle client 19.11.0.0.0, Oracle server 19.11.0.0.0 it will automatically release the statement, so the problem DOESN'T OCCURS. So, the problem is seems not related to AdoDb, but to be dependent on PHP version or OCI8 PHP module version. All others component are the same, also PHP, OCI8 and Oracle configuration.

Just an update:
My system administrator has found that the problem on php 7.4.19 is related to a setting of Zend Server Version: 2021.0.0,
called zray.enable_cli (zray debugger for cli).
If this parameter is set to "1" (true), php doesn't  automatically close cursors that it open in insert or update statement. 
It close all cursors only at the end of the script, when the open cursors have already beyond the "max open cursors" value and so the PHP script is already dead with ORA-01000 exception.

By setting this parameter to "0" the cursors are automatically closed every time and the problem doesn't occur.

> @dregad dregad added the bug label 2 days ago
> @dregad dregad added this to the v5.22.0 milestone 2 days ago

So this is not an AdoDb bug, but probably a Zend Server bug!
For AdoDb is just an improvement in order to have a method to release/free a statement, like oci functions have.

**Comment by dregad** [comment]
@somax80 I added a new releaseStatement() method to the ADOConnection object.if you get a chance to test the code in the PR I just linked and provide feedback, it would be appreciated.

**Comment by somax80** [comment]
> @somax80 I added a new releaseStatement() method to the ADOConnection object.if you get a chance to test the code in the PR I just linked and provide feedback, it would be appreciated.

I've tested the new method and i could confirm that it succesfully release the oci statement referenced in $stmt[1].

Before calling releaseStatement, $stmt[1] value is "resource(...) of type (oci8 statement)"
After calling it, $stmt value become "resource(...) of type (Unknown)". 

Only a suggestion: It is not be better to explicity unset $stmt[1] after calling oci_free_statement, if oci_free_statement return true?

**Comment by dregad** [comment]
Thanks for testing and feedback!

> Only a suggestion: It is not be better to explicity unset $stmt[1] after calling oci_free_statement, if oci_free_statement return true?

Yes that makes sense, good suggestion I will do that.

**Comment by dregad** [comment]
@somax80 I just updated the PR following your suggestion.
